### PR TITLE
feat(locksmith) - Filter items on members page by email address and checked in time 

### PR DESCRIPTION
**Pull request by kalidiagne**
# Description
Add a new endpoint to get keys and filter it with different paraments and get results at once, actually, the frontend do a lot of operation and merge of data. this endpoint manages all from locksmith with a single endpoint 

# Issues

Fixes #
Refs #

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [x] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [x] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: locksmith/src/routes/v2/member.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+const router = express.Router({ mergeParams: true })
+
+const web3Service = new Web3Service(networks)
+const memberController = new MemberController({
+  web3Service,
+})
+
+router.get(
+  '/:network/locks/:lockAddress/list',
```
Comment (julien51) [inline locksmith/src/routes/v2/member.ts:15]:
```suggestion
  '/:network/locks/:lockAddress/members',
```

File: locksmith/src/controllers/v2/memberController.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+   * List of members with additional metadata when caller is the lockManager
+   * @return {members} members list
+   */
+  async list(request: Request, response: Response) {
+    try {
+      const lockAddress = Normalizer.ethereumAddress(request.params.lockAddress)
+      const network = Number(request.params.network)
+      const { filters } = request.query
+
+      const loggedUserAddress = Normalizer.ethereumAddress(
```
Comment (julien51) [inline locksmith/src/controllers/v2/memberController.ts:23]:
```suggestion
      const loggedInUserAddress = Normalizer.ethereumAddress(
```

File: locksmith/src/controllers/v2/memberController.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        request!.user!.walletAddress
+      )
+
+      const isLockOwner = await this.web3Service.isLockManager(
+        lockAddress,
+        loggedUserAddress,
+        network
+      )
+
+      const client = new Members(network)
```
Comment (julien51) [inline locksmith/src/controllers/v2/memberController.ts:33]: Let's move this into a `MemberOperation` file so we can re-use.

File: locksmith/openapi.yml [hunk 4]
```diff
@@ -46,3 +46,39 @@ paths:
       responses:
         '200':
           description: OK
+  /v2/member/{network}/locks/{lockAddress}/list:
```
Comment (julien51) [inline locksmith/openapi.yml:49]:
```suggestion
  /v2/member/{network}/locks/{lockAddress}/members:
```

File: locksmith/__tests__/controllers/v2/memberController.test.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    Web3Service: jest.fn().mockImplementation(() => {
+      return {
+        isLockManager: (lock: string) => lockAddress === lock,
+      }
+    }),
+  }
+})
+
+describe('Member v2 endpoints for locksmith', () => {
+  it('Get members without authorization', async () => {
```
Comment (julien51) [inline locksmith/__tests__/controllers/v2/memberController.test.ts:21]: Please check that no private data is served/added otherwise this test is not useful.

**Review by julien51 — CHANGES_REQUESTED** [review]
Left a few comments.

File: locksmith/__tests__/controllers/v2/memberController.test.ts [hunk 6]
```diff
@@ -18,49 +19,131 @@ jest.mock('@unlock-protocol/unlock-js', () => {
...
-  it('Get members without authorization', async () => {
-    expect.assertions(1)
-    const getListResponse = await request(app).get(
-      `/v2/api/member/${network}/locks/${lockAddress}/members`
-    )
-    expect(getListResponse.status).toBe(403)
-  })
-
-  it('Get members without filter query throw error', async () => {
+  it('should get members without filters throw error', async () => {
```
Comment (julien51) [inline locksmith/__tests__/controllers/v2/memberController.test.ts:22]: I am confused. Can you rephrase please?

File: locksmith/__tests__/controllers/v2/memberController.test.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+    expect(getMembersResponse.status).toBe(200)
+
+    expect(Array.isArray(getMembersResponse.body)).toStrictEqual(true)
+    const [firstItem] = getMembersResponse.body ?? []
+    expect(firstItem).toHaveProperty('token')
+    expect(firstItem).toHaveProperty('lockName')
+    expect(firstItem).toHaveProperty('expiration')
+    expect(firstItem).toHaveProperty('keyholderAddress')
+    expect(firstItem).toHaveProperty('lockAddress')
```
Comment (julien51) [inline locksmith/__tests__/controllers/v2/memberController.test.ts:55]: We should check that results are indeed active (not expired)

**Review by julien51 — COMMENTED** [review]
I worry that the tests are doing _real_ network calls. They should not and only use mocks..

File: locksmith/openapi.yml [hunk 8]
```diff
@@ -46,3 +46,39 @@ paths:
       responses:
         '200':
           description: OK
+  /v2/member/{network}/locks/{lockAddress}/members:
```
Comment (julien51) [inline locksmith/openapi.yml:49]:
```suggestion
  /v2/member/{network}/locks/{lockAddress}/keys:
```

Ok so now I realize this is not listing members, but keys (tokens). Sorry I was confused earlier because you used `list`.

File: locksmith/openapi.yml [hunk 9]
```diff
@@ -46,3 +46,39 @@ paths:
...
+                  type: string
+                filterKey:
+                  type: string
+                expiration:
+                  type: string
+                page: 
+                  type: integer
+      responses:
+        '200':
+          description: List of members
```
Comment (julien51) [inline locksmith/openapi.yml:80]:
```suggestion
          description: List of keys matching the query
```

File: locksmith/src/controllers/v2/memberController.ts [hunk 10]
```diff
@@ -0,0 +1,54 @@
+import { Request, Response } from 'express'
+import Normalizer from '../../utils/normalizer'
+import logger from '../../logger'
+import * as memberOperations from '../../operations/memberOperations'
+
+export default class MemberController {
```
Comment (julien51) [inline locksmith/src/controllers/v2/memberController.ts:6]: Sorry let's rename this to `keyController` ?

File: locksmith/src/graphql/datasource/members.ts [hunk 11]
```diff
@@ -0,0 +1,182 @@
+import { GraphQLDataSource } from 'apollo-datasource-graphql'
+import networks from '@unlock-protocol/networks'
+import gql from 'graphql-tag'
+import { DocumentNode } from 'graphql'
+import { getValidNumber } from '../../utils/normalizer'
+
+export type MemberFilter = 'all' | 'active' | 'expired' | 'keyId'
```
Comment (julien51) [inline locksmith/src/graphql/datasource/members.ts:7]: Same here...

**Comment by kalidiagne** [comment]
Update with more tests and mocks. 
Also changed the naming from `members` -> `keys` for more clarity

File: locksmith/__tests__/controllers/v2/keyController.test.ts [hunk 12]
```diff
@@ -0,0 +1,314 @@
+import request from 'supertest'
+import { loginRandomUser } from '../../test-helpers/utils'
+
+const app = require('../../../src/app')
+
+jest.setTimeout(600000)
```
Comment (julien51) [inline locksmith/__tests__/controllers/v2/keyController.test.ts:6]: Any reason why you need this? It's usually an indication that there are network calls happening and there should not be any :/

File: locksmith/__tests__/controllers/v2/keyController.test.ts [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+    expect(loginResponse.status).toBe(200)
+
+    const getKeysResponse = await request(app)
+      .get(`/v2/api/${network}/locks/${lockAddress}/keys`)
+      .set('authorization', `Bearer ${loginResponse.body.accessToken}`)
+
+    expect(getKeysResponse.status).toBe(404)
+  })
+
+  it('should containts metadata when caller is the lockManager', async () => {
```
Comment (julien51) [inline locksmith/__tests__/controllers/v2/keyController.test.ts:129]:
```suggestion
  it('should contain metadata when caller is the lockManager', async () => {
```

File: locksmith/__tests__/controllers/v2/keyController.test.ts [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+    expect.assertions(5)
+
+    const { loginResponse } = await loginRandomUser(app)
```
Comment (julien51) [inline locksmith/__tests__/controllers/v2/keyController.test.ts:132]: Hum, if the user is a random user, they are likely not the lock manager?

File: locksmith/__tests__/controllers/v2/keyController.test.ts [hunk 15]
```diff
@@ -0,0 +1,314 @@
...
+
+    const [res] = getKeysResponse.body
+    expect(getKeysResponse.status).toBe(200)
+    expect(getKeysResponse.body.length).toBe(1)
+    expect(res.email).toBe('example@gmai.com')
+    expect(res.address).toBe('brescia')
+    expect(res.firstname).toBe('mario rossi')
+  })
+
+  it('should not containts metadata when caller is not lockManager', async () => {
```
Comment (julien51) [inline locksmith/__tests__/controllers/v2/keyController.test.ts:150]:
```suggestion
  it('should not contains metadata when caller is not lockManager', async () => {
```

File: locksmith/__tests__/controllers/v2/keyController.test.ts [hunk 16]
```diff
@@ -0,0 +1,314 @@
...
+    expect.assertions(5)
+
+    const { loginResponse, address } = await loginRandomUser(app)
+    const lockAddress = address
```
Comment (julien51) [inline locksmith/__tests__/controllers/v2/keyController.test.ts:154]: I am confised here, why would the address of the user be the lock address?

**Comment by kalidiagne** [comment]
Close because splitted in smaller PR

**Comment by kalidiagne** [comment]
Close because splitted in smaller PR
